[CON-50] 프로젝트 공고 상세화면 지원하기 바텀시트 작성

robot156 opened the pull request:
## 작업내용
[CON-50] 프로젝트 공고 상세화면 지원하기 바텀시트 작성 
- Fcm Token 추가
- intro 화면에서 알람 권한 요청 추가
- 홈 아이템 변경사항 적용
- BaseAlertDialogFragment 작성
- 다이얼로그 호출 시 바텀메뉴가 사라지는 이슈 수정
- tb_banner -> tab_banner로 xml id 변경
- 신청하기 API 연동
- 신청하기 바텀시트 작성
- 지원사유 다이얼로그 작성

## 스크린샷
| before | after |
:-:|:-:
<img src="https://github.com/Connect-Crew/TeamOne-AOS/assets/66992191/8ab69b3c-a873-4d03-af8a-d98f128e4d29" width="250"/>| <image src="https://github.com/Connect-Crew/TeamOne-AOS/assets/66992191/c5f27cc9-ce46-45d5-be03-20e719d62c35" width="250"/>
<img src="" width="250"/>| <image src="https://github.com/Connect-Crew/TeamOne-AOS/assets/66992191/49636c05-8d5a-4a9a-a8e8-39cf2660da6f" width="250"/>
<img src="" width="250"/>| <image src="https://github.com/Connect-Crew/TeamOne-AOS/assets/66992191/2889608c-11ad-4b27-bc50-38b4b8be8269" width="250"/>
<img src="" width="250"/>| <image src="https://github.com/Connect-Crew/TeamOne-AOS/assets/66992191/c203ab3d-f5e0-4520-822e-e03779db5f9e" width="250"/>
<img src="" width="250"/>| <image src="https://github.com/Connect-Crew/TeamOne-AOS/assets/66992191/47cd4d26-6e07-496d-8ce6-1f0f0a047d2c" width="250"/>
<img src="" width="250"/>| <image src="https://github.com/Connect-Crew/TeamOne-AOS/assets/66992191/7d8d61c2-482a-40a3-9f5c-cda0eb1ef423" width="250"/>
<img src="" width="250"/>| <video src="https://github.com/Connect-Crew/TeamOne-AOS/assets/66992191/ad380962-262c-4f94-8353-f8446a731442" width="250"/>


[CON-50]: https://hyeonjunkang.atlassian.net/browse/CON-50?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/screen/feature/intro/IntroViewModel.kt:
```diff
@@ -26,8 +25,9 @@ class IntroViewModel @Inject constructor(
-    private val currentUserInfo = loadDataSignal.combine(userToken, ::Pair)
-        .flatMapLatest { (_, userToken) -> getUserInfoUseCase(GetUserInfoUseCase.Params(userToken)).asResult() }
+    private val currentUserInfo = userToken
+        .debounce(600)
+        .flatMapLatest { userToken -> getUserInfoUseCase(GetUserInfoUseCase.Params(userToken)).asResult() }
```
이 부분이 FCM 토큰을 받아오는 과정에서 어떤 이슈가 있어서 수정된걸까요?

robot156 replied:
.debounce(600)
위치만 수정된 겁니다~

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/util/firebase/FirebaseUtil.kt:
```diff
@@ -0,0 +1,19 @@
+
+object FirebaseUtil {
+
+    suspend fun getFirebaseMessageToken(): String? = suspendCancellableCoroutine { cancelCoroutine ->
+        Firebase.messaging.token
+            .addOnSuccessListener { cancelCoroutine.resume(it) }
+            .addOnFailureListener {
+                Timber.e("getFirebaseMessageToken exception message [$it]")
+                cancelCoroutine.resume(null)
+            }
+    }
```
addOnFailureListener 가 발생하는 경우가 어떤 경우가 있을까요? 
만약 FailureListener 가 호출되고 유저 fcm 에 null 이 들어가게 된다면 나중에 알람을 오출할 때 문제가 생길 거 같은데, 어떤 방식으로 동작되는지 궁금합니다

robot156 replied:
null 들어가면 서버에서 제어할거에요.
네트워크 이슈나 파이어베이스 이슈 때문에 Fail이 들어올 수 있겠죠

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/model/project/ProjectFeedDetail.kt:
```diff
@@ -27,9 +27,12 @@ data class ProjectFeedDetail(
-    val totalMaxCount: Int,
-    val isEnroll: Boolean = totalCurrentCount < totalMaxCount
-) : Parcelable
+    val totalMaxCount: Int
+) : Parcelable {
+
+    val isEnroll: Boolean
+        get() = totalCurrentCount < totalMaxCount
+}
```
ProjectFeed 랑 공백 형식이 다른 거 같습니다

robot156 replied:
수정해두겠습니다

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/screen/base/BaseAlertDialogFragment.kt:
```diff
@@ -0,0 +1,37 @@
+    protected val dataBinding: T
+        get() = _dataBinding!!
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        return dataBinding.root
+    }
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        _dataBinding = DataBindingUtil.inflate(LayoutInflater.from(requireContext()), layoutResId, null, false)
+
+        return MaterialAlertDialogBuilder(requireContext())
+            .setView(dataBinding.root)
+            .create()
+    }
```
이전 프래그먼트들에선 라이프사이클 순서대로 위치하려고 한다고 느꼈는데, 
이 부분도 onCreateView 랑 onCreateDialog 위치 변경하는건 어떨까요?

robot156 replied:
넵. 수정해두겠습니다.

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/screen/feature/main/home/HomeFragment.kt:
```diff
@@ -36,17 +37,16 @@ import timber.log.Timber
-            homeViewModel.setSelectedCategory(it.category)
-        }
+        HomeCategoryAdapter { homeViewModel.setSelectedCategory(it.category) }
     }
     private val homeProjectFeedAdapter by lazy {
         HomeProjectFeedAdapter(
-            onClickFavoriteProjectFeed = { homeViewModel.setProjectFeedLike(it) },
-            onClickProjectFeed = { homeViewModel.navigateToProjectFeedDetail(it) }
+            onClickFavoriteProjectFeed = homeViewModel::setProjectFeedLike,
+            onClickMemberCount = homeViewModel::navigateToRecruitmentNoticeSummary,
```
어뎁터 사이에 공백이 없는데 어뎁터끼리는 따로 공백을 추가하지 않으려고 하신걸까요?

robot156 replied:
변수끼리는 굳이 간격 띄울 필요가 없을 것 같아서요.

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/screen/feature/main/home/HomeViewModel.kt:
```diff
@@ -61,6 +62,9 @@ class HomeViewModel @Inject constructor(
     private val _navigateToProjectFeedDetail = MutableEventFlow<Int>()
     val navigateToProjectFeedDetail: EventFlow<Int> = _navigateToProjectFeedDetail
 
+    private val _navigateToRecruitmentNoticeSummary = MutableEventFlow<Unit>()
+    val navigateToRecruitmentNoticeSummary: EventFlow<Unit> = _navigateToRecruitmentNoticeSummary
+
```
`nav_home` 의 액션하고 다이얼로그 클래스 이름이 `SummeryRecruitmentNoticeDialog` 인 거 같은데`navigateToSummeryRecruitmentNotice` 으로 통일하는 건 어떨까요??

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/screen/feature/main/home/SummaryRecruitmentNoticeDialog.kt:
```diff
@@ -0,0 +1,62 @@
+        initObserver()
+    }
+
+    private fun initView() {
+        with(dataBinding) {
+            rvMember.apply {
+                adapter = summaryMemberAdapter
+                layoutManager = LinearLayoutManager(requireContext())
+            }
+        }
+    }
```
추가로 설정해야 할 뷰가 따로 없으면, `dataBinding.rvMember.apply { ... }` 으로 변경해주셔도 좋을 거 같습니다

nodobi replied:
모집 현황 옆에 나오는 전체 인원 수에 해당하는 부분이 추가되야 할 거 같아요
![image](https://github.com/Connect-Crew/TeamOne-AOS/assets/76645322/466df76f-efbd-4552-8605-d56e23213b95)

nodobi commented on presentation/src/main/res/layout/dialog_summary_recruitment_notice.xml:
```diff
@@ -0,0 +1,106 @@
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toEndOf="@id/tv_member"
+            app:layout_constraintTop_toTopOf="@id/tv_member"
+            tools:text="3/ 7" />
```
너무 사소한 내용이긴 한데,  3하고 / 사이에 띄어쓰기가 안 되어 있는 거 같아요

nodobi commented on presentation/src/main/res/layout/item_project_feed.xml:
```diff
@@ -141,6 +179,7 @@
                     android:textColor="@color/color_d62246"
                     tools:text="49" />
             </LinearLayout>
+
         </androidx.constraintlayout.widget.ConstraintLayout>
```
공백 남아있습니당

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/adapter/project/enrollment/ProjectEnrollmentAdapter.kt:
```diff
@@ -0,0 +1,67 @@
+        override fun areContentsTheSame(oldItem: RecruitStatus, newItem: RecruitStatus): Boolean {
+            return oldItem == newItem
+        }
+    }
+) {
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProjectEnrollmentViewHolder {
+        val inflater = LayoutInflater.from(parent.context)
```
onCreateViewHolder 앞에 공백이 빠진 거 같습니당

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/screen/base/BaseAlertDialogFragment.kt:
```diff
@@ -18,6 +24,18 @@ abstract class BaseAlertDialogFragment<T : ViewDataBinding> : DialogFragment() {
     protected val dataBinding: T
         get() = _dataBinding!!
 
+    private val loadingDialog: AppCompatDialog by lazy {
+        DialogLoadingBinding.inflate(LayoutInflater.from(requireContext()), null, false)
```
`Fragment` 에서 제공되는 `layoutInflater` 가 있는거로 아는데, 혹시 사용을 안하신 이유가 있을까요?

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/screen/feature/project/projectintroduction/ProjectEnrollmentBottomSheetDialogFragment.kt:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    override fun onDestroyView() {
+        dataBinding.rvProjectEnroll.adapter = null
+        super.onDestroyView()
+    }
```
`_dataBinding` 을 `onDestroyView()` 에서 null 로 바꾸는 이유와 같은 맥락으로 이해했는데 맞을까요?

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/screen/feature/project/projectintroduction/ProjectEnrollmentReasonAlertDialog.kt:
```diff
@@ -0,0 +1,111 @@
+    override fun onPause() {
+        super.onPause()
+        dataBinding.etProjectEnrollmentReason.removeTextChangedListener(projectEnrollReasonTextChangeListener)
+    }
```
이거도 `_dataBinding` 을 `onDestroyView()` 에서 null 로 바꾸는 이유와 같은 맥락으로 이해하면 되는걸까요?

nodobi commented on presentation/src/main/res/layout/dialog_project_enrollment_reason.xml:
```diff
@@ -0,0 +1,94 @@
+        android:orientation="vertical"
+        tools:context=".screen.feature.project.projectintroduction.ProjectEnrollmentReasonAlertDialog">
+
+        <TextView
+            android:id="@+id/tv_title"
+            style="@style/Widget.TeamOne.TextView.Body2"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="20dp"
+            android:text="@{@string/project_detail_enrollment_reason(viewModel.selectedRecruitPart.part)}"
+            tools:text="백엔드 엔지니어에 지원합니다." />
```
피그마랑 텍스트 형식이 조금 다른 거 같아요. 저는 지금 형태가 더 괜찮아보여서 피그마에 언급해주시면 좋을 거 같습니당

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/screen/feature/project/ProjectDetailContainerFragment.kt:
```diff
@@ -102,6 +105,26 @@ class ProjectDetailContainerFragment : BaseFragment<FragmentProjectDetailContain
         }
     }
 
+    private fun initNavBackStackObserve() {
+        val navBackStackEntry = findNavController().getBackStackEntry(R.id.projectDetailContainerFragment)
+
+        val resultObserver = LifecycleEventObserver { _, event ->
+            if (event == Lifecycle.Event.ON_RESUME) {
+                if (!navBackStackEntry.savedStateHandle.contains(KEY_IS_PROJECT_UPDATE)) return@LifecycleEventObserver
+                val isUpdate: Boolean = navBackStackEntry.savedStateHandle[KEY_IS_PROJECT_UPDATE] ?: false
+                if (isUpdate) projectDetailContainerViewModel.invalidateProjectDetail()
+                navBackStackEntry.savedStateHandle.remove<Boolean>(KEY_IS_PROJECT_UPDATE)
+            }
+        }
+
+        navBackStackEntry.getLifecycle().addObserver(resultObserver)
+        viewLifecycleOwner.lifecycle.addObserver(LifecycleEventObserver { _, event ->
+            if (event == Lifecycle.Event.ON_DESTROY) {
+                navBackStackEntry.getLifecycle().removeObserver(resultObserver)
+            }
+        })
+    }
+
```
메소드 이름이 `Observe` 로 끝나는데 의도하신건지 확인 부탁드리겠습니당

또 나중에 유저가 피드를 보고 있을 때 리더가 내용을 변경하는 경우와 같이 프로젝트 디테일에 변경이 이루어지면 업데이트 하려고 추가 하신거같은데 맞을까요?

nodobi replied:
아 지금은 유저가 지원을 했을 때 상태를 업데이트 해주시려는 이유였군요,..!

robot156 replied:
observer로 수정해놓을게요

nodobi commented on presentation/src/main/java/com/connectcrew/presentation/screen/feature/sign/signin/SignInFragment.kt:
```diff
@@ -124,7 +125,7 @@ class SignInFragment : BaseFragment<FragmentSignInBinding>(R.layout.fragment_sig
     }
 
     private fun initGoogle() {
-        val googleClientId = requireContext().getString(R.string.GOOGLE_CLIENT_ID)
+        val googleClientId = BuildConfig.GOOGLE_CLIENT_ID
```
기존 방식과 차이점이 뭔지 궁금해요. 
저번에 말씀해주셨던 보안과 관련된 문제일까요?

robot156 replied:
GOOGLE_CLIENT_ID / GOOGLE_CLIENT_ID_DATA
Key가 두 개 있었는데 지금보니 두 개를 나눌 필요가 없더라구요.
그래서 하나로 통일했습니다

nodobi reviewed (approved):
리뷰 완료!
저도 작업해야 할 부분이라고 생각하니까 궁금한 점이 너무 많았네요.. ㅋㅋㅋ
작업하시느라 고생하셨습니다!

robot156 committed:
[Minjin.Kim] [CON-50] 프로젝트 공고 상세화면 지원하기 바텀시트 작성
- FCM 토큰 설정
- 인트로에서 알람 권한 체크 설정
- 프로젝트 상세 바텀시트 화면 작성
- 홈 화면 ui 변경사항 적용
- BaseAlertDialogFragment / BaseBottomSheetFragment 작성
